### PR TITLE
soc: esp32: linker script updates

### PR DESCRIPTION
**Pull request by sylvioalves**
This PR splits `common-rom.ld` into smaller linker files to enable custom placement of sections. It also updates ESP32 family to make use of this approach.

This is useful for architectures as such as ESP32 that requires log strings to be placed in RAM area 
due to flash cache access. It can also be useful for applications that have small ROM footprint.

### CODE REVIEW

File: include/linker/linker-tool-gcc.h
Comment (sylvioalves) [inline include/linker/linker-tool-gcc.h:114]: @dcpleung @andyross This is the only way I was able to have `common-rom.ld` content properly mapped into ESP virtual memory. Otherwise all those k_objects hardcoded in RODATA section needs to be kept.  
In case this is not the right approach, any recommendation is welcome.

Comment (dcpleung) [reply include/linker/linker-tool-gcc.h:114]: I am worried that this change will affect other XIP enabled platforms since the linking will be done differently.

IIRC, when virtual memory is first introduced, it is not compatible with XIP, and there was never an effort to make them work together (yet). The issue here is that both of these require the same linker magic (like the line here) to work.

And it seems your later commits are disabling CONFIG_XIP, so why do you still need this?

Comment (sylvioalves) [reply include/linker/linker-tool-gcc.h:114]: @dcpleung, Later commits are not disabling XIP. In fact, it is setting enabled by default in Kconfig.soc. `select XIP`.  
I am also afraid about the behavior on other platforms.
I just found this:  
```c
#if Z_MEM_VM_OFFSET != 0
#define Z_VM_KERNEL 1
#ifdef CONFIG_XIP
#error "XIP and a virtual memory kernel are not allowed"
#endif
#endif
```
What would be the alternatives?  
a) Having custom ESP check in that file to enable this?
b) Some different approach in ESP linker script? (I could not find a way for this though).

@nashif Do you have something to add?

Comment (dcpleung) [reply include/linker/linker-tool-gcc.h:114]: One way I can think of is not to enable XIP, and build with virtual memory addresses. However, some XIP-like magic will need to be introduced into your reset code to copy the data and BSS regions into memory. Or if you have a separate bootloader, do the loading there, and Zephyr kernel does not have to worry about copying.

Comment (sylvioalves) [reply include/linker/linker-tool-gcc.h:114]: @dcpleung @andyross , boot up code already does that, but we still need both LMA and VMA, which is missing for current ROM sections (which is intended in this commit). So if boot code is supposed to do that mapping when XIP is enabled, then this change is correct.  
Also, `GROUP_DATA_LINK_IN` has the same check added in this commit:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/linker/linker-tool-gcc.h#L134  
Make sense having `GROUP_ROM_LINK_IN` to have as well.

Comment (dcpleung) [reply include/linker/linker-tool-gcc.h:114]: Since you are using virtual memory, I think it would be better to set `Z_VM_KERNEL`. Change `#ifdef CONFIG_MMU` to `#ifdef CONFIG_KERNEL_VM_SUPPORT` would probably work in `include/sys/mem_manage.h`.

**Comment by sylvioalves** [comment]
Removed snippets commit from this PR to #43725.

File: include/linker/linker-tool-gcc.h
Comment (sylvioalves) [reply include/linker/linker-tool-gcc.h:114]: Well, it did not work using this approach.  
Have you checked my previous comment regarding `GROUP_DATA_LINK_IN` having this XIP check?

Comment (dcpleung) [reply include/linker/linker-tool-gcc.h:114]: `GROUP_DATA_LINK_IN` is for data region which is supposed to be copied to RAM if XIP is enabled, hence the need for load address. RODATA is NOT supposed to be copied to RAM. If you change the #if here, it will definitely affect other XIP boards as now they will have a load address for RODATA where previously not.

To avoid affecting other boards, it is be possible to undefine `GROUP_ROM_LINK_IN` in your linker script so you can redefine it to suit your need.

Comment (sylvioalves) [reply include/linker/linker-tool-gcc.h:114]: But redefining `GROUP_ROM_LINK_IN` in the linker won't handle the case  as in this test, will it?  
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/misc/iterable_sections/sections-rom.ld

File: soc/xtensa/esp32/linker.ld
Comment (sylvioalves) [reply soc/xtensa/esp32/linker.ld:22]: sure, I am updating the PR with a new approach and with necessary comments.

File: subsys/logging/Kconfig.misc
Comment (sylvioalves) [inline subsys/logging/Kconfig.misc:74]: @nordic-krch @dcpleung PTAL this approach so that we can have custom log* sections placement.

File: include/linker/linker-tool-gcc.h
Comment (sylvioalves) [reply include/linker/linker-tool-gcc.h]: Closing this conversation as this approach isn't valid.

**Review by uLipe — DISMISSED** [review]

File: subsys/logging/Kconfig.misc
Comment (sylvioalves) [reply subsys/logging/Kconfig.misc:74]: It works either way for me. Who else can suggest or approve this (or yours)?

Comment (dcpleung) [reply subsys/logging/Kconfig.misc:74]: @nordic-krch may have some thoughts about this...

**Comment by sylvioalves** [comment]
Only rebased to solve conflicts.  

@dcpleung @cfriedt @mbolivar-nordic, PTAL.

**Comment by cfriedt** [comment]
Maybe @nordic-krch as well?

File: include/linker/common-rom.ld
Comment (mbolivar-nordic) [inline include/linker/common-rom.ld:175]: Have you thought about how and whether this should integrate with the armclang linker generation support? I don't see any discussion of this in the commit message or PR description.

**Review by mbolivar-nordic — COMMENTED** [review]
Most of the changes in here are in soc-specific areas where I'm happy to defer to you. I do think that the change to the common-rom.ld file needs some more thought and explanation. I won't block this PR, but I don't feel comfortable approving either unless those questions are answered a bit more.

File: include/linker/common-rom.ld
Comment (sylvioalves) [reply include/linker/common-rom.ld:175]: @mbolivar-nordic, I am not aware of how armclang linker generation works. What is the link between that and `common-rom.ld`?  I'll be glad to bring this to discussion if any.  

`CONFIG_LOG_CUSTOM_SECTION` is default n, which won't make any difference to platforms using `common-rom.ld`.

**Comment by sylvioalves** [comment]
> Maybe @nordic-krch as well?

If you check the whole PR, @nordic-krch review was requested 2 weeks ago.

**Comment by dcpleung** [comment]
I think it would be better to break down `common-rom.ld` into different files, each one having a snippet of the same subsys. `common-rom.ld` will include all these. And if SoCs/boards need to override this, they simply don't include `common-rom.ld` but individual snippets so these snippets can be placed at desired location. This avoids the need for a separate kconfig.

**Comment by dcpleung** [comment]
Can be divided into `kernel`, `cpp`, `network`, `bluetooth`, `log`, and `misc` (for smaller ones that don't really need to be extracted).

**Comment by dcpleung** [comment]
I did a quick pass on splitting `common-rom.ld` @ https://github.com/zephyrproject-rtos/zephyr/pull/44709

File: subsys/logging/Kconfig.misc
Comment (sylvioalves) [reply subsys/logging/Kconfig.misc]: Did not go on with this approach.

File: include/linker/common-rom.ld
Comment (sylvioalves) [reply include/linker/common-rom.ld]: @mbolivar-nordic closing this thread as I did not follow this approach.

**Comment by sylvioalves** [comment]
@dcpleung Added your suggestion/PR.

**Comment by dcpleung** [comment]
LGTM

**Review by uLipe — DISMISSED** [review]

**Review by glaubermaroto — DISMISSED** [review]

**Comment by sylvioalves** [comment]
Rebased to add latest BT_IAS config.

**Review by mbolivar-nordic — DISMISSED** [review]
This looks reasonable to me.

**Review by glaubermaroto — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
Needs rebase

**Comment by sylvioalves** [comment]
Rebased and re-organized `device-handles.ld`. @ceolin, PTAL.

**Review by glaubermaroto — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**
